### PR TITLE
Append the labels that don't exist in metricStringKeys

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
We should append the labels that don't exist in metricStringKeys

Failed to start datacatalog in single binary because we attempted to register the same metric key.

Error message:
```
 Failed to register metrics. Error: descriptor Desc{fqName: \"datacatalog:datacatalog:storage:read_failure\", help: 
\"Indicates failure in GET for a given reference\", constLabels: {}, variableLabels: [app_name project domain exec_id wf 
node task tasktype runtime_type runtime_version failure_type failure_type]} is invalid: duplicate label names [goroutine 83 
[running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 
+0x65\ngithub.com/flyteorg/datacatalog/pkg/rpc/datacatalogservice.NewDataCatalogService.func
```

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2266

## Follow-up issue
_NA_

### CODE REVIEW

File: promutils/labeled/counter.go [hunk 1]
```diff
@@ -59,8 +68,15 @@ func NewCounter(name, description string, scope promutils.Scope, opts ...MetricO
...
 			c.Counter = scope.MustNewCounter(GetUnlabeledMetricName(name), description)
 		} else if additionalLabels, casted := opt.(AdditionalLabelsOption); casted {
-			c.CounterVec = scope.MustNewCounterVec(name, description, append(metricStringKeys, additionalLabels.Labels...)...)
-			c.additionalLabels = contextutils.MetricKeysFromStrings(additionalLabels.Labels)
+			var labels []string
+			for _, label := range additionalLabels.Labels {
+				if contains(metricStringKeys, label) == false {
+					labels = append(labels, label)
+				}
+			}
```
Comment (EngHabu) [inline promutils/labeled/counter.go:76]: Nit:
```suggestion
			labels := make([]string, 0, len(additionalLabels.Labels)
			metricKeysSet := sets.NewString(metricStringKeys...)
			for _, label := range additionalLabels.Labels {
				if !metricKeysSet.Contains(label) {
					labels = append(labels, label)
				}
			}
```

Comment (EngHabu) [reply promutils/labeled/counter.go:76]: And then we can delete the contains function

Comment (pingsutw) [reply promutils/labeled/counter.go]: updated it, thank you @EngHabu

**Review by EngHabu — APPROVED** [review]

File: promutils/labeled/counter.go [hunk 2]
```diff
@@ -59,8 +61,16 @@ func NewCounter(name, description string, scope promutils.Scope, opts ...MetricO
...
+			labels := make([]string, 0, len(additionalLabels.Labels))
+			metricKeysSet := sets.NewString(metricStringKeys...)
+			for _, label := range additionalLabels.Labels {
+				if !metricKeysSet.Has(label) {
+					labels = append(labels, label)
+				}
+			}
+			// Here we only append the labels that don't exist in metricStringKeys
+			c.CounterVec = scope.MustNewCounterVec(name, description, append(metricStringKeys, labels...)...)
+			c.additionalLabels = contextutils.MetricKeysFromStrings(labels)
```
Comment (EngHabu) [inline promutils/labeled/counter.go:73]: Should we apply the same logic to other metrics under labeled package?

Comment (pingsutw) [reply promutils/labeled/counter.go:79]: yeah, updated it

**Review by EngHabu — APPROVED** [review]
Thank you!

File: promutils/labeled/counter.go [hunk 3]
```diff
@@ -59,8 +73,10 @@ func NewCounter(name, description string, scope promutils.Scope, opts ...MetricO
 		if _, emitUnlabeledMetric := opt.(EmitUnlabeledMetricOption); emitUnlabeledMetric {
 			c.Counter = scope.MustNewCounter(GetUnlabeledMetricName(name), description)
 		} else if additionalLabels, casted := opt.(AdditionalLabelsOption); casted {
-			c.CounterVec = scope.MustNewCounterVec(name, description, append(metricStringKeys, additionalLabels.Labels...)...)
-			c.additionalLabels = contextutils.MetricKeysFromStrings(additionalLabels.Labels)
+			labels := GetUniqueLabels(metricStringKeys, additionalLabels.Labels)
+			// Here we only append the labels that don't exist in metricStringKeys
+			c.CounterVec = scope.MustNewCounterVec(name, description, append(metricStringKeys, labels...)...)
```
Comment (honnix) [inline promutils/labeled/counter.go:78]: This `append` here may modify the same underly array backing up `metricStringKeys` and cause corruption for a counter that was created previously.

A repro:

```go
package labeled

import (
	"context"
	"testing"

	"github.com/flyteorg/flytestdlib/contextutils"
	"github.com/flyteorg/flytestdlib/promutils"
	"github.com/stretchr/testify/assert"
)

func TestLabeledCounter(t *testing.T) {
	UnsetMetricKeys()
	assert.NotPanics(t, func() {
		SetMetricKeys(contextutils.ProjectKey, contextutils.DomainKey, contextutils.WorkflowIDKey)
	})

	scope := promutils.NewTestScope()
	c1 := NewCounter(
		"handle_count1",
		"c1",
		/*scope=*/ scope,
		/*opts=*/ AdditionalLabelsOption{Labels: []string{"foo"}},
	)
	c2 := NewCounter(
		"handle_count2",
		"c2",
		/*scope=*/ scope,
		/*opts=*/ AdditionalLabelsOption{Labels: []string{"tasktype"}}, // this corrupts c1
	)

	ctx := context.TODO()
	ctx = contextutils.WithProjectDomain(ctx, "project", "domain")
	ctx = contextutils.WithWorkflowID(ctx, "workflow")
	ctx = contextutils.WithTaskID(ctx, "task")
	ctx = contextutils.WithTaskType(ctx, "type")
	ctx = context.WithValue(ctx, "foo", "foo")

	c1.Inc(ctx) // <--- panic here
	c2.Inc(ctx)
}
```

cc @hamersaw

Comment (honnix) [reply promutils/labeled/counter.go:78]: It was not introduced by this PR though, the last change of this was ages ago in https://github.com/pingsutw/flytestdlib/commit/c811acb9176b93f55fce4f8a6f3d0339cee0f9de. :)
